### PR TITLE
Added ability to set an alternative transferrer

### PR DESCRIPTION
**Pull request by yahgwai**
Provides the ability to set an alternative transferrer address that can transfer express lane controller rights for a specific express lane controller.

After chatting with @gzneoth, I think it makes sense to set a transferrer per express lane controller rather than a transferrer per bid. If we put the transferrer optionally in the bid, then it would require extra storage and call data when people did set that option.

### CODE REVIEW

**Review by edfelten — CHANGES_REQUESTED** [review]
The spec should specify who has transfer rights after the alternate transferrer does a transfer.

It's probably best to say that the alternate transferrer can only transfer the rights once, and if it does a transfer, the alternate transferrer address is set to zero (or whatever represents a null value).

**Comment by yahgwai** [comment]
> The spec should specify who has transfer rights after the alternate transferrer does a transfer.
> 
> It's probably best to say that the alternate transferrer can only transfer the rights once, and if it does a transfer, the alternate transferrer address is set to zero (or whatever represents a null value).

Why would the alternate transferrer be only able to transfer rights once? I was envisaging a mapping(expressLaneController => alternativeTransferrer) that an express lane controller would set once, and then only update it if they wanted to use a different transferrer.

**Comment by yahgwai** [comment]
Also, I want to clarify which of the following is correct for when the rights can be transferred:
1. From the moment the auction is resolved in round r to the end of round r
2. From the moment the auction is resolved in round r to the end of round r+1.
I had assumed 2. was correct as it allows the express lane controller to transfer during any time they control their round, and any time before it after they have won.

**Comment by edfelten** [comment]
> > The spec should specify who has transfer rights after the alternate transferrer does a transfer.
> > It's probably best to say that the alternate transferrer can only transfer the rights once, and if it does a transfer, the alternate transferrer address is set to zero (or whatever represents a null value).
> 
> Why would the alternate transferrer be only able to transfer rights once? I was envisaging a mapping(expressLaneController => alternativeTransferrer) that an express lane controller would set once, and then only update it if they wanted to use a different transferrer.

My thought was that the party receiving the rights wants to know that they won't be rugged by another transfer. 

Perhaps we should leave this issue open by having the transfer rights operation take an argument saying whether to allow further transfers. Then smart contract based rights markets would have the choice of how to do it.

**Comment by edfelten** [comment]
> Also, I want to clarify which of the following is correct for when the rights can be transferred:
> 
> 1. From the moment the auction is resolved in round r to the end of round r
> 2. From the moment the auction is resolved in round r to the end of round r+1.
>    I had assumed 2. was correct as it allows the express lane controller to transfer during any time they control their round, and any time before it after they have won.

Yes, 2 is what was intended.

**Comment by yahgwai** [comment]
> My thought was that the party receiving the rights wants to know that they won't be rugged by another transfer.
> 
> Perhaps we should leave this issue open by having the transfer rights operation take an argument saying whether to allow further transfers. Then smart contract based rights markets would have the choice of how to do it.

Ah ok, so the point is to stop a last minute change of alternative transferrer, only allowing a new transferrer to be set if the alternative transferrer is set to 0? This could be problematic if no transfer actually takes place, since then the transferrer would never be reset.

A similar mechanism to achieve the same thing would be for the controller to provide a round number when setting the alternative transferrer, once that round number is reached then the transferrer could be changed but not before. A round number of 0 could be used for controllers who want to be able to transfer at any time

File: research_spec.md [hunk 1]
```diff
@@ -36,7 +36,7 @@ For a round that will start its active phase at time `t`, the round proceeds lik
...
...ignation of the express lane controller for the round starting at t and the end of the round (t+`RoundDurationSeconds`), the express lane controller, or an alternative `transferrer` if set, can call the `transferExpressLaneController` on the auction contract to transfer the express lane controller right for the round to a new address. By calling `transferExpressLaneController` and transferring express lane controller rights the current controller loses all of its own rights in the round. The transferee receives express lane controller rights from the moment they are transferred, as well as the right to transfer to another address, either directly or by an alternative transferrer if set. Each express lane controller can set their own alternative `transferrer` by calling `setTransferrer` function at any time, this `transferrer` will have the ability to transfer the rights of any current or future round controlled by that express lane controller. Along with the new transferrer address, an optional round number is provided when calling `setTransferrer` which specifies when the transferrer can next be updated. If a round number is not provided the transferrer can be updated at any time.
```
Comment (terencechain) [inline research_spec.md]: > round starting at t and the end of the round (t+`RoundDurationSeconds`)

Shouldn't this be `t - AuctionClosingSeconds` and `t + RoundDurationSeconds`

Comment (yahgwai) [reply research_spec.md]: We cant give transferral rights until we know who the express lane controller is - which may happen at any time between t-`AuctionClosingSeconds` and t

Comment (terencechain) [reply research_spec.md]: Oh ya. I meant to say when the auction is resolved to t + RoundDurationSeconds. I'm not sure how to phrase it though

Comment (yahgwai) [reply research_spec.md]: Yeh, clearer to write "auction resolved" than "designation of express lane controller". I'll update.

**Review by edfelten — COMMENTED** [review]
One question here is how nonces / sequence numbers are supposed to work if there is a change of controll.er

**Comment by yahgwai** [comment]
> One question here is how nonces / sequence numbers are supposed to work if there is a change of controll.er

Updated to specify that the sequencer will reset the round sequence number to 0 upon a transfer.

File: research_spec.md [hunk 2]
```diff
@@ -36,7 +36,9 @@ For a round that will start its active phase at time `t`, the round proceeds lik
...
+  
+  Each express lane controller can set their own alternative `transferor` by calling `setTransferor` function at any time, this `transferor` will have the ability to transfer the rights of any current or future round controlled by that express lane controller. Along with the new transferor address, an optional round number is provided when calling `setTransferor` which specifies when the transferor can next be updated. If a round number is not provided the transferor can be updated at any time.
```
Comment (edfelten) [inline research_spec.md]: Presumably the transferror for round `i` can't say that the transferror can next be updated in round `i+100`, right?  Doesn't there need to be some limit on this, so a controller can't block transfers by a later controller?

**Review by edfelten — CHANGES_REQUESTED** [review]
One question to address

File: research_spec.md (hunk 2 above)
Comment (yahgwai) [reply research_spec.md]: Each bidder can set their own transferror. So the limit they impose is only on themselves, not on any other controller.

Comment (edfelten) [reply research_spec.md]: Yes, that makes sense. But I think the spec should say it explicitly.

Comment (yahgwai) [reply research_spec.md]: Updated

**Review by edfelten — APPROVED** [review]
Looks good

**Pull request merged**